File: Lambda/android/src/main/java/com/amazonaws/reactnative/lambda/AWSRNLambdaClient.java
//
// Copyright 2010-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License").
// You may not use this file except in compliance with the License.
// A copy of the License is located at
//
// http://aws.amazon.com/apache2.0
//
// or in the "license" file accompanying this file. This file is distributed
// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing
// permissions and limitations under the License.
//

package com.amazonaws.reactnative.lambda;

import com.amazonaws.reactnative.core.AWSRNClientConfiguration;
import com.amazonaws.reactnative.core.AWSRNClientMarshaller;
import com.amazonaws.reactnative.core.AWSRNCognitoCredentials;
import com.amazonaws.reactnative.core.BackgroundRunner;
import com.amazonaws.reactnative.core.BackgroundRunner.Supplier;
import com.amazonaws.regions.Region;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.lambda.*;
import com.amazonaws.services.lambda.model.*;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.json.JSONObject;

import java.nio.ByteBuffer;

public class AWSRNLambdaClient extends ReactContextBaseJavaModule {

    private AWSLambdaClient lambdaClient;
    private Gson gson;

    private final BackgroundRunner backgroundRunner = new BackgroundRunner();

    public AWSRNLambdaClient(final ReactApplicationContext context) {
        super(context);
    }

    @Override
    public String getName() {
        return "AWSRNLambdaClient";
    }

    @ReactMethod
    public void initWithOptions(final ReadableMap options) {
        if (!options.hasKey("region")) {
            throw new IllegalArgumentException("expected region key");
        }
        final AWSRNCognitoCredentials credentials = this.getReactApplicationContext().getNativeModule(AWSRNCognitoCredentials.class);
        if (credentials.getCredentialsProvider() == null) {
            throw new IllegalArgumentException("AWSCognitoCredentials is not initialized");
        }
        gson = new GsonBuilder().setFieldNamingStrategy(FieldNamingPolicy.UPPER_CAMEL_CASE).registerTypeAdapter(ByteBuffer.class, AWSRNClientMarshaller.getSerializer()).registerTypeAdapter(ByteBuffer.class, AWSRNClientMarshaller.getDeserializer()).create();
        lambdaClient = new AWSLambdaClient(credentials.getCredentialsProvider(), new AWSRNClientConfiguration().withUserAgent("Lambda"));
        lambdaClient.setRegion(Region.getRegion(Regions.fromName(options.getString("region"))));
    }

    @ReactMethod
    public void Invoke(final ReadableMap options, final Promise promise) {
        backgroundRunner.runInBackground(
            new Supplier<WritableMap>() {
                @Override
                public WritableMap get() throws Exception {
                    final InvokeRequest request = gson.fromJson(
                        new JSONObject(AWSRNClientMarshaller.readableMapToMap(options)).toString(),
                        InvokeRequest.class
                    );
                    final InvokeResult response = lambdaClient.invoke(request);
                    return AWSRNClientMarshaller.jsonToReact(new JSONObject(gson.toJson(response)));
                }
            },
            promise
        );
    }

}

File: Lambda/ios/Lambda/AWSRNLambdaClient.m
//
// Copyright 2010-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the &quot;License&quot;).
// You may not use this file except in compliance with the License.
// A copy of the License is located at
//
// http://aws.amazon.com/apache2.0
//
// or in the &quot;license&quot; file accompanying this file. This file is distributed
// on an &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing
// permissions and limitations under the License.
//
#import "AWSRNLambdaClient.h"
#import "AWSRNHelper.h"
#import "AWSRNCognitoCredentials.h"


@implementation AWSRNLambdaClient{
    AWSLambda* Lambda;
}
@synthesize bridge = _bridge;

RCT_EXPORT_MODULE(AWSRNLambdaClient);

RCT_EXPORT_METHOD(initWithOptions:(NSDictionary *)options){
    AWSRNHelper *helper = [[AWSRNHelper alloc]init];
    AWSRNCognitoCredentials* CognitoCredentials = (AWSRNCognitoCredentials*)[_bridge moduleForName:@"AWSRNCognitoCredentials"];
    int count = 0;
    while(![CognitoCredentials getCredentialsProvider] && count != 5){
        [NSThread sleepForTimeInterval:0.25];
        count++;
    }
    if(![CognitoCredentials getCredentialsProvider]){
        @throw [NSException exceptionWithName:@"InvalidArgument" reason:@"AWSCognitoCredentials is not initialized" userInfo:options];
    }
    AWSServiceConfiguration *configuration = [[AWSServiceConfiguration alloc] initWithRegion:[helper regionTypeFromString:[options objectForKey:@"region"]] credentialsProvider:[CognitoCredentials getCredentialsProvider]];
    [AWSLambda registerLambdaWithConfiguration:configuration forKey:@"react-native-Lambda"];
    [configuration addUserAgentProductToken:@"AWSLambda"];
    self.Lambda = [AWSLambda LambdaForKey:@"react-native-Lambda"];
}


RCT_EXPORT_METHOD(Invoke:(NSDictionary*)options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    NSError *error;
    AWSLambdaInvocationRequest *request = [AWSMTLJSONAdapter modelOfClass:[AWSLambdaInvocationRequest class] fromJSONDictionary:options error:&error];
    if(error){
        @throw [NSException exceptionWithName:@"InvalidArgument" reason:error.localizedDescription userInfo:error.userInfo];
    }
    [[self.Lambda invoke:request] continueWithBlock:^id _Nullable(AWSTask * _Nonnull task) {
        if(task.error){
            reject([NSString stringWithFormat:@"%ld",task.error.code], [NSString stringWithFormat:@"%@",task.error.userInfo], task.error);
        }else{
            if([task.result isKindOfClass:[AWSModel class]]){
                NSDictionary *dict = [AWSMTLJSONAdapter JSONDictionaryFromModel:task.result];
                resolve(dict);
            }else{
                resolve(@{}) ;
            }
        }
        return nil;
    }];
}


@end
